Verify contributions of future round aggregations

hrxi opened the pull request:
Fixes #1434.

hrxi committed:
Remove unused function `ValidatorNetwork::cache`

hrxi committed:
Let the request/response protocol handle the connecting

Previously, this had to be emulated on the validator network level.

hrxi committed:
Verify contributions of future round aggregations

Fixes #1434.

nibhar commented on tendermint/src/tendermint.rs:
```diff
@@ -390,52 +391,104 @@ impl<TProtocol: Protocol> Tendermint<TProtocol> {
+                        if round > self.state.current_round {
+                            // future contributions need to be tracked separately, as once a round has f+1 contributions
+                            // it can be fast tracked
+                            let contributors = self.future_contributions.entry(round).or_default();
+                            *contributors |= new_contributors;
```
Maybe consider sticking with the Entry API here. 
```suggestion
                            self.future_contributions.entry(round)
                                .and_modify(|contributors| *contributors |= new_contributors)
                                .or_insert(new_contributors);
```

hrxi replied:
Same ownership problem, but here, I also actually like the current version better.

hrxi committed:
Check for unverified future round proposals after reading new messages

nibhar commented on tendermint/src/tendermint.rs:
```diff
@@ -390,52 +391,104 @@ impl<TProtocol: Protocol> Tendermint<TProtocol> {
+
+            // If no future round aggregation verification is currently running, start a new one.
+            if self.future_round_verification.is_none() && !self.future_round_messages.is_empty() {
+                // TODO: choose at random
```
Or choose with weight. Not every contribution is equally good.|

Picking the first one is likely not vey good as that will (in all nodes) prefer low index validators over others. While weighting the contributions is maybe not necessary I think random selection is.

hrxi replied:
Now chosen at random.

nibhar commented on tendermint/src/tendermint.rs:
```diff
@@ -390,52 +391,104 @@ impl<TProtocol: Protocol> Tendermint<TProtocol> {
+                let id = message.tag;
+                if id.0 > self.state.current_round {
+                    // Queue the message for future round aggregation verification.
+                    match self
+                        .future_round_messages
+                        .entry(message.aggregation.sender())
+                    {
+                        Entry::Vacant(v) => {
+                            v.insert(message);
+                        }
+                        Entry::Occupied(mut o) => {
+                            // Generally: Newer is better. If the new message comes
+                            // from the same round as a previous one, assume that
+                            // it is better.
+                            if id.0 >= o.get().tag.0 {
+                                o.insert(message);
+                            }
+                        }
                     }
```
I like the Entry API better here as well.
```suggestion
                    match self
                        .future_round_messages
                        .entry(message.aggregation.sender())
                        .and_modify(|m| {
                            if id.0 > m.tag.0 {
                                *m = message
                            }
                        })
                        .or_insert(message)
```

hrxi replied:
I also like that better, but it's not possible due to ownership rules, unfortunately. The closure would take ownership of `message`, but it's needed afterwards.

nibhar commented on validator-network/src/network_impl.rs:
```diff
@@ -43,37 +45,30 @@ where
-            .await
-            .map_err(|_| NetworkError::Unreachable)?
+    /// Returns the local validator ID.
+    fn local_validator_id(&self) -> u16 {
```
This will panic for non elected, but active validators.

hrxi replied:
Made it a `Result` instead.

nibhar commented on validator-network/src/network_impl.rs:
```diff
@@ -178,60 +196,18 @@ where
         validator_id: usize,
         msg: M,
```
`M` likely no longer needs to be `Clone`. Can't mention it there, but it is 3 lines up.

hrxi replied:
Done.

nibhar commented on validator-network/src/network_impl.rs:
```diff
@@ -178,60 +196,18 @@ where
+        }
     }
 
     async fn request<TRequest: Request>(
```
With `request` and `send_to` now being virtually the same, we may want to look into unifying them into one.

hrxi replied:
Not so easy because of the different trait requirements.

nibhar replied:
Just looked a bit more into it and seams not worth the effort if even worth it at all.

nibhar reviewed (changes requested):
Looks good! Mostly minor comments, except for the selection of the next verification task. I think randomness there is important, as the selection otherwise is based on a value for which the entire network has the same bias.

hrxi committed:
Choose a future round aggregation for verification at random

hrxi committed:
Don't panic in `ValidatorNetwork::send_to` if not an elected validator

hrxi committed:
Make `#[derive(SerializeContent)]` work with lifetime parameters

hrxi committed:
Fix historic transaction hash collisions

Include block number and network ID in the hash. Also make sure that
they can't clash with normal transactions.

Fixes #1935.

nibhar reviewed (approved):
LGTM
